File: examples/Microsoft.Spark.CSharp.Examples/Sql/Batch/Basic.cs
// Licensed to the .NET Foundation under one or more agreements.
// The .NET Foundation licenses this file to you under the MIT license.
// See the LICENSE file in the project root for more information.

using System;
using System.Collections.Generic;
using Microsoft.Spark.Sql;
using Microsoft.Spark.Sql.Types;
using static Microsoft.Spark.Sql.Functions;

namespace Microsoft.Spark.Examples.Sql.Batch
{
    /// <summary>
    /// A simple example demonstrating basic Spark SQL features.
    /// </summary>
    internal sealed class Basic : IExample
    {
        public void Run(string[] args)
        {
            if (args.Length != 1)
            {
                Console.Error.WriteLine(
                    "Usage: Basic <path to SPARK_HOME/examples/src/main/resources/people.json>");
                Environment.Exit(1);
            }

            SparkSession spark = SparkSession
                .Builder()
                .AppName("SQL basic example using .NET for Apache Spark")
                .Config("spark.some.config.option", "some-value")
                .GetOrCreate();

            // Need to explicitly specify the schema since pickling vs. arrow formatting
            // will return different types. Pickling will turn longs into ints if the values fit.
            // Same as the "age INT, name STRING" DDL-format string.
            var inputSchema = new StructType(new[]
            {
                new StructField("age", new IntegerType()),
                new StructField("name", new StringType())
            });
            DataFrame df = spark.Read().Schema(inputSchema).Json(args[0]);

            Spark.Sql.Types.StructType schema = df.Schema();
            Console.WriteLine(schema.SimpleString);

            IEnumerable<Row> rows = df.Collect();
            foreach (Row row in rows)
            {
                Console.WriteLine(row);
            }

            df.Show();

            df.PrintSchema();

            df.Select("name", "age", "age", "name").Show();

            df.Select(df["name"], df["age"] + 1).Show();

            df.Filter(df["age"] > 21).Show();

            df.GroupBy("age")
                .Agg(Avg(df["age"]), Avg(df["age"]), CountDistinct(df["age"], df["age"]))
                .Show();

            df.CreateOrReplaceTempView("people");

            // Registering Udf for SQL expression.
            DataFrame sqlDf = spark.Sql("SELECT * FROM people");
            sqlDf.Show();

            spark.Udf().Register<int?, string, string>(
                "my_udf",
                (age, name) => name + " with " + ((age.HasValue) ? age.Value.ToString() : "null"));

            sqlDf = spark.Sql("SELECT my_udf(*) FROM people");
            sqlDf.Show();

            // Using UDF via data frames.
            Func<Column, Column, Column> addition = Udf<int?, string, string>(
                (age, name) => name + " is " + (age.HasValue ? age.Value + 10 : 0));
            df.Select(addition(df["age"], df["name"])).Show();

            // Chaining example:
            Func<Column, Column> addition2 = Udf<string, string>(str => $"hello {str}!");
            df.Select(addition2(addition(df["age"], df["name"]))).Show();

            // Multiple UDF example:
            df.Select(addition(df["age"], df["name"]), addition2(df["name"])).Show();

            // UDF return type as array.
            Func<Column, Column> udfArray =
                Udf<string, string[]>((str) => new[] { str, str + str });
            df.Select(Explode(udfArray(df["name"]))).Show();

            // UDF return type as map.
            Func<Column, Column> udfMap =
                Udf<string, IDictionary<string, string[]>>(
                    (str) => new Dictionary<string, string[]> { { str, new[] { str, str } } });
            df.Select(udfMap(df["name"]).As("UdfMap")).Show(truncate: 50);

            // Joins.
            DataFrame joinedDf = df.Join(df, "name");
            joinedDf.Show();

            DataFrame joinedDf2 = df.Join(df, new[] { "name", "age" });
            joinedDf2.Show();

            DataFrame joinedDf3 = df.Join(df, df["name"] == df["name"], "outer");
            joinedDf3.Show();
            
            // Union of two data frames
            DataFrame unionDf = df.Union(df);
            unionDf.Show();

            // Add new column to data frame
            df.WithColumn("location", Lit("Seattle")).Show();

            // Rename existing column
            df.WithColumnRenamed("name", "fullname").Show();

            // Filter rows with null age
            df.Filter(Col("age").IsNull()).Show();

            // Fill null values in age column with -1
            df.Na().Fill(-1, new[] { "age" }).Show();

            // Drop age column
            df.Drop(new[] { "age" }).Show();

            spark.Stop();
        }
    }
}

File: examples/Microsoft.Spark.FSharp.Examples/Sql/Basic.fs
// Licensed to the .NET Foundation under one or more agreements.
// The .NET Foundation licenses this file to you under the MIT license.
// See the LICENSE file in the project root for more information.

namespace Microsoft.Spark.Examples.Sql

open System
open Microsoft.Spark.Examples
open Microsoft.Spark.Sql

type Basic() =
    member this.Run(args : string[]) =
        match args with
        | [| filePath |] ->
            let spark = SparkSession.Builder().AppName("Hello F#").GetOrCreate()

            let df = spark.Read().Json(filePath)

            let schema = df.Schema()
            printfn "%s" (schema.SimpleString)

            for row in df.Collect() do
                printfn "%s" (row.ToString())
        
            df.Show()

            df.PrintSchema()

            df.Select("name", "age", "age", "name").Show()

            df.Select(df.["name"], df.["age"] + 1).Show()

            df.Filter(df.["age"].Gt(21)).Show()
        
            df.GroupBy("age")
                .Agg(Functions.Avg(df.["age"]),
                        Functions.Avg(df.["age"]),
                        Functions.CountDistinct(df.["age"], df.["age"]))
                        .Show()

            // SQL example.
            df.CreateOrReplaceTempView("people")

            // Registering UDF for SQL expression.
            let sqlDf = spark.Sql("SELECT * FROM people")
            sqlDf.Show()

            spark.Udf().Register<Nullable<int>, string, string>(
                "my_udf",
                fun age name ->
                    name + " with " + (if age.HasValue then (string)(age.Value) else "null"))

            let sqlDf = spark.Sql("SELECT my_udf(*) FROM people")
            sqlDf.Show()

            // Using UDF via data frames.
            let addition = Functions.Udf<Nullable<int>, string, string>(
                            fun age name ->
                                name + " is " +
                                    (if age.HasValue then (string)(age.Value + 10) else "0"))

            df.Select(addition.Invoke(df.["age"], df.["name"])).Show()

            // Chaining example:
            let addition2 = Functions.Udf<string, string>(fun str -> "hello " + str + "!")
            df.Select(addition2.Invoke(addition.Invoke(df.["age"], df.["name"]))).Show()

            // Multiple UDF example:
            df.Select(addition.Invoke(df.["age"], df.["name"]), addition2.Invoke(df.["name"]))
                .Show()

            // Joins.
            let joinedDf = df.Join(df, "name")
            joinedDf.Show()

            let joinedDf2 = df.Join(df, ["name"; "age"] |> List.toSeq)
            joinedDf2.Show()

            let joinedDf3 = df.Join(df, df.["name"].EqualTo(df.["name"]), "outer")
            joinedDf3.Show()
            
            // Union of two data frames
            let unionDf = df.Union(df)
            unionDf.Show()

            // Add new column to data frame
            df.WithColumn("location", Functions.Lit("Seattle")).Show()

            // Rename existing column
            df.WithColumnRenamed("name", "fullname").Show()

            // Filter rows with null age
            df.Filter(df.["age"].IsNull()).Show()

            // Fill null values in age column with -1
            df.Na().Fill(-1L, ["age"]).Show()

            // Drop age column
            df.Drop(df.["age"]).Show()

            spark.Stop()
            0
        | _ ->
            printfn "Usage: Basic <path to SPARK_HOME/examples/src/main/resources/people.json>"
            1

    interface IExample with
        member this.Run (args) = this.Run (args)
